src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,552 @@
+// Copyright 2024 TerseTS Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of Swing and Slide Filter algorithms from the paper:
+//! Hazem Elmeleegy, Ahmed K. Elmagarmid, Emmanuel Cecchet, Walid G. Aref, and Willy Zwaenepoel.
+//! Online piece-wise linear approximation of numerical streams with precision guarantees.
+//! Proc. VLDB Endow. 2, 1, 2009.
+//! https://doi.org/10.14778/1687627.1687645.
+
+const std = @import("std");
+const ts = @import("../tersets.zig");
+
+const mem = std.mem;
+const testing = std.testing;
+
+const ArrayList = std.ArrayList;
+
+const Segment = struct {
+    start_time: usize,
+    start_value: f64,
+    end_time: usize,
+    end_value: f64,
+};
+
+/// Line structure uses f80 to avoid overflows with f64
+const Line = struct {
+    slope: f80,
+    intercept: f80,
+};
+
+pub fn isWithinErrorBound(
```

Comment:
I think we should move that function at the end of tersets.zig, allowing any other compression method to have access to it.

Reply:
Fixed. Now all public functions are at the beginning after the structures. Probably we should make `appendValueAndIndexToArrayList` private in `src/functional/poor_mans_compression.zig`.

Reply:
I agree `appendValueAndIndexToArrayList()` should be private. It is unclear to me why private functions are not placed after the last public function that uses them in `swing_slide_filter.zig` as suggested above. For example, `appendValue()` seems to be placed after `decompress()` despite only being used by `compress_swing()`. Also the order of the private functions is unclear, e.g, why are `usizeToF80()` placed above all the functions that use it and `appendValue()` is placed before `computeSlopeDerivate()` despite `computeSlopeDerivate()`'s first use being before `appendValue()`. It is also fine for me if we use a different ordering of the function then I suggested above, I would just prefer that we are consistent so functions are as easy to find as possible.

Reply:
I see, I did not read your comment carefully and put all private functions at the end. Somehow, I thought that the public functions, compress and decompress, should be at the beginning since they are the most important for the users. In fact, I think I like this structure so you agree we can keep it like that. I can then reorganize the private functions to be in the same order that are called in the file. What do you think?

Reply:
Sounds fine to me. So just to be sure we agree, we organize files as types, public functions in the order they are supposed to be called, private functions in the order they are called? If so, this should probably be stated in the README.md. Maybe add a task to the TODO list to create a development section with information like this in the README.md.

Reply:
Yes, while going through the file was very convenient to have compress and decompress functions first. The private functions will be added as they are called.